File: chapter1-Hanoi-tower/solutions/chapter1_Colin6618.js
function hanoi(level, sourcePillar, transportPillar, targetPillar, callback) {
  if(level === 1) callback(sourcePillar, targetPillar);
  else {
    hanoi(level - 1, sourcePillar, targetPillar, transportPillar, callback);
    hanoi(1, sourcePillar, transportPillar, targetPillar,  callback);
    hanoi(level - 1, transportPillar, sourcePillar, targetPillar, callback);
  }
}

function getCounter() {
  var count = 1;
  return function(pillarA, pillarB) {
    console.log(`${count}: ${pillarA} => ${pillarB}`);
    count++;
  }
}

function play(level, sourcePillar, transportPillar, targetPillar) {
  var counter = getCounter();
  hanoi(level, sourcePillar, transportPillar, targetPillar, counter);
}

module.exports = play;

// test
// node -e 'require("./chapter1_Colin6618.js")(3, "left", "middle", "right")'

File: chapter1-Hanoi-tower/solutions/chapter1_Colin6618.py
def hanoi(level, sourcePillar, transportPillar, targetPillar, callback):
    if level == 1:
        callback(sourcePillar, targetPillar)
    else:
        hanoi(level - 1, sourcePillar, targetPillar, transportPillar, callback)
        hanoi(1, sourcePillar, transportPillar, targetPillar,  callback)
        hanoi(level - 1, transportPillar, sourcePillar, targetPillar, callback)

def getCounter():
    a = [1]
    def count(pillarA, pillarB):
        print (a[0], pillarA, pillarB)
        a[0] = a[0] + 1
        return a[0]
    return count


def play(level, sourcePillar, transportPillar, targetPillar):
    counter = getCounter()
    hanoi(level, sourcePillar, transportPillar, targetPillar, counter)

play(5, "left", "middle", "right");
